runtime/lua/vim/lsp/_watchfiles.lua

```diff
@@ -0,0 +1,486 @@
+      watched_paths[path].callbacks[client_id][reg_id].callback(
+        path,
+        protocol.FileChangeType.Created
+      )
+    end
+  end
+
+  if recursive_watch then
+    return
+  end
+
+  local scan, err, errname = uv.fs_scandir(path)
+  if errname == 'ENOTDIR' or errname == 'ENOENT' then
+    return
+  end
+  assert(not err, err)
+  while true do
+    local ret = { uv.fs_scandir_next(scan) }
+    if #ret == 0 then
+      break
+    end
+    if #ret == 3 then
+      assert(not ret[2], ret[2])
+    end -- error check
+    local name, type = ret[1], ret[2]
+    if watch_each_file or type == 'directory' then
+      local subpath = filepath_join(path, name)
+      local include = true
+      for _, exclude in ipairs(excludes) do
+        if M._match(exclude, subpath) then
+          include = false
+          break
+        end
+      end
+      if not include then
+        break
+      end
+      fsevent_ensure_recursive(subpath, pattern, kind, client_id, reg_id, type)
+    end
+  end
+end
+
+local function ensure_registrations()
+  for client_id, regs in pairs(registrations) do
+    for reg_id, watchers in pairs(regs) do
+      for _, watcher in ipairs(watchers) do
+        fsevent_ensure_recursive(
+          watcher.base_dir,
+          watcher.pattern,
+          watcher.kind,
+          client_id,
+          reg_id,
+          'directory'
+        )
+      end
+    end
+  end
+end
+
+--- Registers the workspace/didChangeWatchedFiles capability dynamically.
+---
+---@param reg table LSP Registration object.
+---@param ctx table Context from the |lsp-handler|.
+function M.register(reg, ctx)
+  local client = vim.lsp.get_client_by_id(ctx.client_id)
+  local watchers = {}
+  for _, w in ipairs(reg.registerOptions.watchers) do
+    local glob_patterns = {}
+    if type(w.globPattern) == 'string' then
+      for _, folder in ipairs(client.workspace_folders) do
+        table.insert(glob_patterns, { baseUri = folder.uri, pattern = w.globPattern })
+      end
+    else
+      table.insert(glob_patterns, w.globPattern)
+    end
+    for _, glob_pattern in ipairs(glob_patterns) do
+      local pattern = parse(glob_pattern.pattern)
+      local base_dir = nil
+      if type(glob_pattern.baseUri) == 'string' then
+        base_dir = glob_pattern.baseUri
+      elseif type(glob_pattern.baseUri) == 'table' then
+        base_dir = glob_pattern.baseUri.uri
+      end
+      assert(base_dir, "couldn't identify root of watch")
+      base_dir = vim.uri_to_fname(base_dir)
+      path_sep = path_sep or base_dir:match('^([a-zA-Z]:)(.*)') and '\\' or '/'
+      local kind = w.kind
+        or protocol.WatchKind.Create + protocol.WatchKind.Change + protocol.WatchKind.Delete
+
+      table.insert(watchers, {
+        base_dir = base_dir,
+        pattern = pattern,
+        kind = kind,
+      })
+    end
+  end
+  registrations[ctx.client_id] = registrations[ctx.client_id] or {}
+  registrations[ctx.client_id][reg.id] = watchers
+  ensure_registrations()
```

Comment:
Could this loop over the watchers only?

```suggestion
  ensure_registrations()
  for _, watcher in ipairs(watchers) do
    fsevent_ensure_recursive(
            watcher.base_dir,
            watcher.pattern,
            watcher.kind,
            ctx.client_id,
            reg.id,
            'directory'
          )
  end
```


There's afaik no need here to re-process other clients/registrations

Reply:
Yeah I don't think I'll have to do this anymore with the new changes.